mypy/ipc.py

```diff
@@ -47,36 +47,63 @@ class IPCBase:
 
     connection = None  # type: _IPCHandle
 
-    def __init__(self, name: str) -> None:
+    def __init__(self, name: str, timeout: Optional[float]) -> None:
         self.READ_SIZE = 100000
         self.name = name
+        if sys.platform == 'win32':
+            # Windows timeouts are in miliseconds
+            self.timeout = int(timeout * 1000) if timeout is not None else None
+        else:
+            self.timeout = timeout
 
     def read(self) -> bytes:
         """Read bytes from an IPC connection until its empty."""
         bdata = bytearray()
-        while True:
-            if sys.platform == 'win32':
-                more, _ = _winapi.ReadFile(self.connection, self.READ_SIZE)
-            else:
+        if sys.platform == 'win32':
+            while True:
+                ov, err = _winapi.ReadFile(self.connection, self.READ_SIZE, overlapped=True)
+                try:
+                    if err == _winapi.ERROR_IO_PENDING:
+                        timeout = self.timeout if self.timeout else _winapi.INFINITE
+                        res = _winapi.WaitForSingleObject(ov.event, timeout)
+                        assert res == _winapi.WAIT_OBJECT_0
+                except BaseException:
+                    ov.cancel()
+                    raise
+                finally:
```

Comment:
I've not fully reviewed this, but one initial thing: I don't immediately understand the logic here of why this is in a finally block?

Reply:
Yeah I agree it is weird, this is an artifact of the multiprocessing code. I think it should be okay to remove it.